federatedrec/factorization_machine/hetero_factorization_machine/hetero_fm_arbiter.py

```diff
@@ -0,0 +1,142 @@
+            self.one_vs_rest_fit(train_data=data_instances, validate_data=validate_data)
+        else:
+            self.need_one_vs_rest = False
+            self.fit_binary(data_instances, validate_data)
+
+    def fit_binary(self, data_instances=None, validate_data=None):
+        """
+        Train FM model of role arbiter
+        Parameters
+        ----------
+        data_instances: DTable of Instance, input data
+        """
+
+        LOGGER.info("Enter hetero fm model arbiter fit")
+
+        self.cipher_operator = self.cipher.paillier_keygen(self.model_param.encrypt_param.key_length)
+        self.batch_generator.initialize_batch_generator()
+        validation_strategy = self.init_validation_strategy()
+
+        while self.n_iter_ < self.max_iter:
+            iter_loss = None
+            batch_data_generator = self.batch_generator.generate_batch_data()
+            total_gradient = None
+            self.optimizer.set_iters(self.n_iter_)
+            for batch_index in batch_data_generator:
+                # Compute and Transfer gradient info
+                gradient = self.gradient_loss_operator.compute_gradient_procedure(self.cipher_operator,
+                                                                                  self.optimizer,
+                                                                                  self.n_iter_,
+                                                                                  batch_index)
+                if total_gradient is None:
+                    total_gradient = gradient
+                else:
+                    total_gradient = total_gradient + gradient
+                # training_info = {"iteration": self.n_iter_, "batch_index": batch_index}
+                # self.perform_subtasks(**training_info)
+
+                loss_list = self.gradient_loss_operator.compute_loss(self.cipher_operator, self.n_iter_, batch_index)
+
+                if len(loss_list) == 1:
+                    if iter_loss is None:
+                        iter_loss = loss_list[0]
+                    else:
+                        iter_loss += loss_list[0]
+
+            # if converge
+            if iter_loss is not None:
+                iter_loss /= self.batch_generator.batch_num
+                if not self.in_one_vs_rest:
+                    self.callback_loss(self.n_iter_, iter_loss)
+
+            if self.model_param.early_stop == 'weight_diff':
+                weight_diff = fate_operator.norm(total_gradient)
+                LOGGER.info("iter: {}, weight_diff:{}, is_converged: {}".format(self.n_iter_,
+                                                                                weight_diff, self.is_converged))
+                if weight_diff < self.model_param.tol:
+                    self.is_converged = True
+            else:
+                if iter_loss is None:
+                    raise ValueError("Multiple host situation, loss early stop function is not available."
```

Comment:
A little confusion about this line, HeteroFM does not support multi-host, but  but changes "early stop" to weight diff will support multi-host?

Reply:
Removed this check. Thanks.